### PR TITLE
For building linux-aarch64 target for supporting aarch64 effortless pattern

### PR DESCRIPTION
**Pull request by smandaka**
For building linux-aarch64 target for testing effortless pattern for linux-aarch64 target.

## Description
Initially we were getting build error for linux-aarch64 target, after adding "fix_interpreter "${pkg_prefix}/vendor/bin/*" "$_chef_client_ruby" ruby" build was successful. But still getting runtime error while running "chef -version". So then we added line "set_runtime_env LD_LIBRARY_PATH "$LD_RUN_PATH" "

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:

- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have run the pre-merge tests locally and they pass.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### CODE REVIEW

File: habitat/plan.sh [hunk 1]
```diff
@@ -57,9 +58,7 @@ do_verify() {
 
 do_setup_environment() {
   push_runtime_env GEM_PATH "${pkg_prefix}/vendor"
-
   set_runtime_env APPBUNDLER_ALLOW_RVM "true" # prevent appbundler from clearing out the carefully constructed runtime GEM_PATH
-  set_runtime_env SSL_CERT_FILE "$(pkg_path_for cacerts)/ssl/cert.pem"
```
Comment (tpowell-progress) [inline habitat/plan.sh:62]: I see the export to `SSL_CERT_FILE` on line 71, but have we confirmed that this is redundant?

File: habitat/plan.sh [hunk 2]
```diff
@@ -70,6 +69,8 @@ do_prepare() {
   export OPENSSL_INCLUDE_DIR="$(pkg_path_for openssl)/include"
   export SSL_CERT_FILE="$(pkg_path_for cacerts)/ssl/cert.pem"
   export CPPFLAGS="${CPPFLAGS} ${CFLAGS}"
+  set_runtime_env LD_LIBRARY_PATH "$LD_RUN_PATH"
```
Comment (tpowell-progress) [inline habitat/plan.sh:72]: This and line 101 seem fairly generic changes that don't explain how this is implementing the stated goal of the PR headline.

**Review by tpowell-progress — CHANGES_REQUESTED** [review]

**Comment by tpowell-progress** [comment]
Please link to any habitat updates required for this change and fill out the PR description area above to describe how this accomplishes the stated goal.

File: habitat/plan.sh (hunk 1 above)
Comment (mwrock) [reply habitat/plan.sh:62]: Yeah, this remove it's runtime usage and line 71 is used at build time. I'm not sure we want to remove this unless we know for sure we do not need it.

File: habitat/plan.sh (hunk 2 above)
Comment (mwrock) [reply habitat/plan.sh:72]: I know there is more detailed info in the JIRA ticket which I cannot seem to find. @smandaka could you provide the link?

**Review by mwrock — CHANGES_REQUESTED** [review]

File: habitat/plan.sh (hunk 2 above)
Comment (smandaka) [reply habitat/plan.sh:72]: https://chefio.atlassian.net/browse/CHEF-660

**Comment by rahulgoel1** [comment]
@tpowell-progress @mwrock  Could someone pls help understand the errors which are showing up in chef-main-verify pipeline above ?  The team is able to build and run infra-client package locally

**Comment by mwrock** [comment]
@rahulgoel1 at first glance this looks non trivial. I see lots of failures from different infra resources. Some are GLIBC linking issues with various libraries (like python) others are more mysterious.

File: habitat/plan.sh [hunk 3]
```diff
@@ -97,6 +98,7 @@ do_build() {
     build_line "Installing gems from git repos properly ..."
     ruby ./post-bundle-install.rb
     build_line "Installing this project's gems ..."
+    fix_interpreter "${pkg_prefix}/vendor/bin/*" "$_chef_client_ruby" ruby
```
Comment (mwrock) [inline habitat/plan.sh:101]: @tpowell-progress see https://chefio.atlassian.net/browse/CHEF-660?focusedCommentId=82405 for an explanation for `fix_interpreter`

File: habitat/plan.sh (hunk 2 above)
Comment (mwrock) [reply habitat/plan.sh:72]: See https://chefio.atlassian.net/browse/CHEF-660?focusedCommentId=82561 for an explanation of `LD_LIBRARY_PATH`. The linked Jira comment had a different solution using `bundle config --local build.ffi "--with-ldflags=-Wl,-rpath=${LD_RUN_PATH}"` which only addresses the native ffi gem. The `LD_LIBRARY_PATH` accomplishes the same goal but could apply to any native gem.

File: habitat/plan.sh (hunk 1 above)
Comment (rahulgoel1) [reply habitat/plan.sh:62]: @smandaka  could you pls try to bring that change back ( not remove the SSL_CERT_FILE) line from plan file and then try ?

**Review by rahulgoel1 — COMMENTED** [review]
@smandaka  @mwrock  has indicated in the https://chefio.atlassian.net/browse/CHEF-660?focusedCommentId=82405 that he removed fix_interpreter line in the do_install in plan file. 
I don't think that change is there in the plan file here. Could you also bring that change ?

File: habitat/plan.sh (hunk 3 above)
Comment (mwrock) [reply habitat/plan.sh:102]: @smandaka I think you should remove the `fix_interpreter` on line 111

**Review by mwrock — CHANGES_REQUESTED** [review]

File: habitat/plan.sh (hunk 2 above)
Comment (rahulgoel1) [reply habitat/plan.sh:73]: Sounds good Matt. @smandaka  pls baseline the plan to match what matt had originally done. Subsequent to that we can incrementally bring in the changes one by one recommended by Matt and Johny and iterate on build cycle incrementally.

**Comment by rahulgoel1** [comment]
@mwrock @tpowell-progress Finally, we are able to get the build green. There is a linting error on 'rpath'. Could you pls recommend what should be done to resolve that ?

**Review by mwrock — APPROVED** [review]

**Comment by tpowell-progress** [comment]
Holding off on this until we can validate why the macOS kitchen tests are failing

**Comment by rahulgoel1** [comment]
@tpowell-progress  Did you get a chance to review that test failure ? Could you pls let us know if the PR can be merged?

**Comment by jaymzh** [comment]
So, based on the CHEF-660 ticket in JIRA, it seems the REAL problem is that there were two `ld-wrapper.sh`'s, the main one here: https://github.com/habitat-sh/core-plans/blob/main/binutils/ld-wrapper.sh#L56 and then another one that no longer exists here: https://github.com/habitat-sh/core-packages/blob/main/packages/development/tools/misc/binutils/base/ld-wrapper.sh

The former properly jams the contents of `$LD_RUN_PATH` into the `-rpath` args for the linker, but the latter does not.

Doing the re-build of FFI happens to properly build the binary for ARM, but only as a side-effect and would only fix FFI... fixing the linking in the first place is the right fix.

However, since those two files have been collapsed now, it seems like this was maybe fixed elsewhere, though that ENTIRE REPO is gone (or hidden from both @johnmccrae  and I). But probably this was fixed already.

Which means the rebuild of FFI here _shouldn't_ be needed, we think.

The `fix_interpreter` move however is needed for reasons explained elsewhere in this PR.

Also, https://github.com/chef/chef/pull/13769 appears to be a dupe of this.

**Comment by atrniv** [comment]
Hi @jaymzh,

You've made a great point regarding the root cause analysis. As you noted, repairing the linking of the linker may seem like the correct fix, but it isn't quite straightforward. I've gone ahead and created a separate issue [here](https://github.com/habitat-sh/core-packages/issues/56) where I've elaborated more on the complexities of this problem.

The previous behavior of the linker script did a good job at masking the underlying issue of dynamic linking at runtime. But, it also introduces an overhead that's borne by every ELF binary/library in every package.

So, we're faced with a bit of a dilemma. We could alter the way impacted packages like this one are built, or add overhead to every other package for a solution that doesn't quite solve the problem of dynamic linking at runtime. It's a tough call, and I'd appreciate any thoughts or input you might have once you have gone through the linked issue.

**Comment by jaymzh** [comment]
https://github.com/habitat-sh/core-packages/ isn't public, so I cannot see it. Happy to discuss this one I can see the discussion :)

**Review by tpowell-progress — APPROVED** [review]

**Review by jaymzh — CHANGES_REQUESTED** [review]
OK, so the contents of https://github.com/habitat-sh/core-packages/issues/56 is available [here](https://gist.github.com/tpowell-progress/5bd0b803299c7da559b95a782193f8b4) for anyone interested. The second ld-wrapper.sh is also now available [here](https://gist.github.com/tpowell-progress/3bf196eba8f8728c115fac19ef829120).

I want to start with a few things from that habitat issue:

> Scenario 2: Suppose an executable, built in Package A, requires a library from Package B, and both Packages A and B are included in Package C. In that case, it becomes challenging for Package A's executable to locate Package B's library.

I'm not sure what you mean by this. The Linux linker as well as the Mac and Windows linkers, all handle recursive dependencies properly. For example, if I `ldd` firefox, which looks recursively:

```shell
$  ldd /usr/lib/firefox/firefox-bin 
	linux-vdso.so.1 (0x00007ffeca9e5000)
	libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f9cdd200000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f9cdd59f000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f9cdd01e000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f9cdd698000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f9cdd4c0000)
```

You'll see that it picks up libm.so.6, which is NOT in it's direct needs:

```shell
$ objdump -p /usr/lib/firefox/firefox-bin | grep NEEDED
  NEEDED               libstdc++.so.6
  NEEDED               libgcc_s.so.1
  NEEDED               libc.so.6
  NEEDED               ld-linux-x86-64.so.2
```

This sorta of thing is inherent to almost any binary, at least in a Linux system.

Are you saying that when you're setting up the RPATH, you don't yet know where the paths to C will be or something? I'm not clear the problem you're trying to illustrate here.

This part seems interesting though:

> An example of Scenario 1 is chef/chef-infra-client, which functioned with the old linker script as libffi was built within the package, and all runtime dependency rpaths were blindly embedded into the libffi ELF. However, it would malfunction if the libffi from core/libffi was utilized.

So, to clarify here, you're saying there's two libffi objects, one built from a core plan and one built from the chef-infra-client plan, and they have different `rpath`s in each? So the idea is that forcing a build of one in chef-infra-client ala [this PR](https://github.com/chef/chef/pull/13769) forces us to use one built based on the chef-infra-client env overwriting the one from the core build env, yes?

You suggest two possible long-term solutions, one is:

> We could modify the linker to load a library from a cache located at a path defined by an environment variable (e.g., HAB_LD_CACHE). This cache would be located in Package C, and the environment variable would be part of Package C's runtime environment. However, this solution is applicable only on Linux, where the linker can be patched.

What do you mean by "modify the linker"? You mean adding it to `ld.so.conf`? Or some sort of LD_PRELOAD hook? Both of those are far-reaching in ways that seem bad.

That said, the ld-wrapper from core-plans seems to [do precisely this](https://gist.github.com/tpowell-progress/3bf196eba8f8728c115fac19ef829120):

```shell
HAB_DEBUG=${HAB_DEBUG:-${HAB_LD_DEBUG:-${HAB_@env_prefix@_DEBUG:-0}}} \
HAB_LD_RUN_PATH=${HAB_@env_prefix@_LD_RUN_PATH:-${HAB_LD_RUN_PATH:-${LD_RUN_PATH:-""}}} \
@wrapper@ @program@ "$@"
```

> The most practical cross-platform solution, in my opinion, is to have the plan author consciously use LD_LIBRARY_PATH (Linux), DYLD_LIBRARY_PATH (macOS), and PATH (Windows) to handle this scenario correctly on an application-specific basis. I acknowledge that this approach introduces an overhead for the plan author, requiring them to understand which dependency libraries their application dynamically loads at runtime. They would also have to carefully consider scenarios where their packaged binary might invoke a non-Habitat binaries and ensure that any libraries (like glibc) used by those binaries are not overridden as a result.

Here I'm assuming you mean having some shell wrapper that sets up the relevant env variables before calling the binary, ala `firefox` (which is a shell script that does that before calling `firefox-bin`).

This is pretty dangerous in the case of Habitat-built packages, because many of those packages call other non-habitat-built binaries, which might inherit this library path.

Finally, to the comment above:

> The previous behavior of the linker script did a good job at masking the underlying issue of dynamic linking at runtime. But, it also introduces an overhead that's borne by every ELF binary/library in every package.

I disagree with this assessment. It wasn't _masking_ dynamic linking at runtime. Binary `rpath`s are the expected correct way to tell the dynamic linker where to find a binary's libraries at runtime.

It seems to me we should be able to craft exactly the correct rpath. Habitat's core-selling-feature is that it knows the full dependency tree of anything you want to build (the example in all of the early talks was "openssl has a security vuln, you trigger an update of that, and habitat can recursively figure out everything that needs a rebuild, rebuild it and deploy it"). So we should be able to add those - and only those - paths to the LD_RUN_PATH (and thus the binary's `rpath`).

As to the performance issue, when we were adding include paths to ruby and it was checking every single include path for every single gem on every single load, it was definitely causing a significant performance issue because Ruby's loader is incredibly naive (it'll check all paths, even if it finds the gem early on, for example). However, I don't think the unix linker is so naive.

Are we really hitting performance issues from having long rpaths? I mean we're fixing issues in chef that are saving upwards of 85 seconds in runtime, so I'd be wary of over-optimizing for potential performacne issues that were trivial (I don't know if it is, I imagine you have some numbers).

Phew, that was long. Sorry if I misunderstood anything or missed anything.

Can we make sure we keep the discussion here (or somewhere public) until the other repo is open (expected tomorrow or so, AIUI), and make sure to provide gists of any code/issues/etc. referenced via a gist?

So, something like this PR would probably be an OK stop-gap if we absolutely have to get something out the door to fix this ASAP. But @tpowell-progress said he's got lots of other blockers on 18, so lets take the time to try to figure out the right fix.

Finally, @tpowell-progress said he's actively working on ffi issues and is very weary of this changing during that work.

**Comment by atrniv** [comment]
Hi @jaymzh, thank you for going through the description of the problem I put up. However it seems I have caused confusion between dynamic linking and dynamic loading due to poor usage of terms in the issue description. It seems I also mixed up the scenario that the solutions proposed applied to. The solutions proposed are all for scenario 2, not scenario 1. I've corrected these mistakes. I have updated the issue to be more precise and clear in terms of the difference between dynamic linking and dynamic loading, please take another look at the [issue](https://github.com/habitat-sh/core-packages/issues/56). 

The first part you describe about dynamic linking is absolutely right. RPATH / RUNPATH  are the correct solution for dynamic linking. They can correctly handle indirect library dependencies as well as you describe. However the issue I was refering to was not *dynamic linking*, but *dynamic loading*, which is what is happening here with libarchive.

I made another mistake in my assessment of what exactly is happening here. *The plan is not building it's own version of libffi*. It is building a language extension library (vendor/gems/ffi-1.15.5/ext/ffi_c/ffi_c.so). This language extension library in the `ffi` gem dynamically links with `libffi`. The `ffi-libarchive` gem then uses the `ffi` gem to dynamically load `libarchive`. Since the old wrapper embedded the libarchive library directory in the language extension libary's RUNPATH that is built, it worked.

The solution 2 proposed in the issue is to actually patch the source code of the linker and recompile it to change it's standard behaviour. The guix folks have [patched](https://git.savannah.gnu.org/cgit/guix.git/tree/gnu/packages/patches/glibc-dl-cache.patch?id=0236013cd0fc86ff4a042885c735e3f36a7f5c25) their linker to modify the [default lookup order](https://man7.org/linux/man-pages/man8/ld.so.8.html) of DT_NEEDED entries. The default order assuming only RUNPATH and not RPATH (because RPATH is deprecated), is:
- LD_LIBRARY_PATH
- Each directory in the RUNPATH
- The linker cache at `/etc/ld.so.cache`
- Default system directories `/lib`, `/usr/lib`, etc

The guix folks made it:
- LD_LIBRARY_PATH
- The linker cache at a package specific location `/gnu/store/*-<PACKAGE_NAME>/etc/ld.so.cache`
- Each directory in the RUNPATH
- Default system directories `/lib`, `/usr/lib`, etc

The way normal linux distributions optimize the lookup is libraries is by:
- Not setting RPATH or RUNPATH by default
- Installing libraries to `/lib` or `/usr/lib`. 
- Adding a cache entry in `/etc/ld.so.cache` to associate the libary name directly with library path.

In the standard linux distributions this works out because the cache is checked before the system directories and most packages do not use RPATH or RUNPATH.

In light of these clarifications, I stand by my initial assessment that the older linker script behavior is not really a solution to dynamic loading. It worked here because the bundler compiles the language extension library (vendor/gems/ffi-1.15.5/ext/ffi_c/ffi_c.so). In other languages like NodeJS some packages simply download pre-built binaries for each architecture. In such cases there is no opportunity to modify the RUNPATH.

In standard operating systems dynamic loading works because the location of the libraries are known global values (/lib, /usr/lib). This is the Achilles' heel of the design of Habitat, which it shares with its source of inspiration (Nix).

However in the spirit of maintaining backward compatibility and not causing mysterious runtime failures due to dynamic loading for existing plans, I have made the linker script behaviour modifiable. The default behaviour is compatible with the old script, but you can use the new behaviour by setting an environment variable `HAB_LD_LINK_MODE=minimal`. This might be good enough to resolve this issue for now here, but again, it's not really a general solution to dynamic loading within Habitat.

**Comment by atrniv** [comment]
Hi @jaymzh , please take a look a look at Solution 4 that I have outlined in the [issue](https://github.com/habitat-sh/core-packages/issues/56). It was triggered by your comment about preloading libraries. I realized since we control the entire compilation process we can directly override `dlopen` / `LoadLibrary` in a transparent manner. It is more powerful than preloading because it would also work if the C library got statically linked into the end binary. It would also work if there was some security restriction preventing preloading. 

I think it should be a pretty general and powerful approach to solving the dynamic loading problem in Habitat. Let me know your thoughts.

**Comment by jaymzh** [comment]
@atrniv - thanks for all that context and updates.

I'll start with the most prevalent piece:

> However in the spirit of maintaining backward compatibility and not causing mysterious runtime failures due to dynamic loading for existing plans, I have made the linker script behaviour modifiable. The default behaviour is compatible with the old script, but you can use the new behaviour by setting an environment variable HAB_LD_LINK_MODE=minimal. This might be good enough to resolve this issue for now here, but again, it's not really a general solution to dynamic loading within Habitat.

This seems like the right solution - at least in the short term. I believe this because:
1. It allows everything to work by default, even for dynamically loaded (`dlopen`) libraries.
2. For those who *need* different behavior, they use "positive handoff", as we say, to turn off the "default works" behavior, thereby acknowledging they have taken responsibility to set the appropriate dependencies in their Hab Plan to ensure it'll work.

This feels like the right mix to me.

Addressing your Scenario 2 though:

> Scenario 2: Suppose an executable, built in Package A, requires to dynamically load a library via FFI from Package B, and both Packages A and B are dependencies of Package C

In this case, A still requires B. Just like in scenario 1.  Therefore, B not being a runtime dependency of A would be a bug. If A's dependencies are correct ,we can setup the RUNPATH correctly at build time.

Your statement here:

> This means the ELF RUNPATH for all binaries in Package A and B have already been compiled and cannot be modified. In this state, it becomes challenging for Package A's executable to dynamically load Package B's library via a dlopen (Linux/MacOS) or LoadLibrary (Windows) call, as there is no RUNPATH entry in the Package A binary that references Package B.

implies a big in the Hab build project to me. It *sounds* like you're saying that C triggers a build of A and B independently, and As dependencies aren't considered. But instead, if you have a graph of:
```
C --+---> A -> B
    |----------^
```

Then the build system should realize B must be built first, then A can be built with the appropriate dependency of B, and then C can be built.

This problem exists in base OS packages as well. Package C can depend on A which can `dlopen` something from B. But in such a case B *must* be a runtime dependency of A.

Onto your big scary question:

> We can override the dlopen / LoadLibrary function when compiling the application with our own implementation during the compilation process. This can be done without patching / modifying the default system C library code using the following approach:

Yikes. Yes, that would work. However, it's a solution fraught with peril.

Nix - as an entire Distro/OS is tracking libc closely, all the time. They control when it gets updated and can ensure any overrides they do if it are always compatible. They have a whole *team* dedicated to just this.

Chef/Progress has a habitat team, but not a team dedicated to tracking libc and packaging it and working with upstream.

We have *many* examples of "clever" stuff like this that no one knows how it works anymore, or is otherwise very opaque to folks who can't update when it's breaking things. For example, no one knows why Expeditor is creating BuildKite tests in *Blocked* status whenever FFI appears to be anywhere in the title, description, branch or repo name. Wut. Many of the low-level ffi or win32 gems we've written are effectively abandonwear that the rest of the company hard relies on.

Further, it's basically impossible to know that dlopen() has been overridden when you're trying to debug some very subtle issue, 2 years from now, after everyone has stopped looking at all this and can't figure out why ffi-yajl seems broken (again). That's not actually true, you can see it if you know what you're looking for, but it's certainly not obvious if you're not looking for it.

As someone who has had to override dlopen, open, close, and even write... it's a rough place to be. I'd avoid it at all costs unless it's absolutely necessary.

Let me know if I miss-understood anything or missed any questions.

**Comment by atrniv** [comment]
@jaymzh, great, I think the default behavior of the new linker wrapper should be a good enough solution for now. 

However, I would like to clarify that keeping the old default behavior is only a fix for Scenario 1. It does not work with all forms of dynamic loading. 

I think the understanding of Scenario 2 is still not clear. 
- I'm not saying that B is a dependency of A that is not being considered and built correctly. 
- I'm saying that B would never reasonably ever become a dependency of A. This is not a bug in the build system that causes it to fail to recognize the dependency. 
- This because the dependency between A and B is created by their usage in Package C and not inherent within Package A and B.

I've made it concrete by providing an [example repo](https://github.com/atrniv/hab-dynamic-loading) so that it is easier to grasp how it could happen. In this example repo:
- Package A is language runtime: `core/node`
- Package B is a library with some useful functionality: `core/libarchive`
- Package C is an application built with the language: `core/hab-dynamic-loading`

There is no general reason why `core/libarchive` should be either a run or build dependency of `core/node`. Hence there is no way that the path to `core/libarchive` is put into the RUNPATH of the node binary. In this particular example the ultimate `dlopen` call comes from within a pre-built library on linux x86_64 `./node_modules/koffi/build/linux_x64/koffi.node` that has no RUNPATH and so will lookup the RUNPATH of the parent ELF binary (node). This parent ELF binary (node) would never be built in Package C's plan. While this is a very contrived example, it may not be uncommon in certain kinds of applications which make use of plugin based architectures (eg: Browsers, Game Engines, etc). 

It could also be that the actual `dlopen` call is directly inside the Package A and not something that is downloaded during Package C's build. For example in the case of the [Nim language fix](https://github.com/nim-lang/Nim/issues/15194) that I mention in the issue, the actual `dlopen` calls are [within the Nim runtime itself](https://github.com/nim-lang/Nim/blob/825a0e7df40100d92e87fc421a135a84b94f818f/lib/pure/dynlib.nim) that links to the system C library and not provided by some 3rd party package.

In most languages (such as Ruby) this may not be a problem with the old linker script as the language extension library is compiled by the packager within the plan. So there is an opportunity for the linker to stuff the library directories of the Package C's deps (Package A and B) into the RUNPATH of language extension library (`vendor/gems/ffi-1.15.5/ext/ffi_c/ffi_c.so`) . These kind of plans all fall under Scenario 1.

**While Scenario 2 may not be something we encounter commonly today, it still exists. The problem of dynamic loading from non-standard locations still exists and does not go away with the old linker wrapper behavior**

After some trial and error I have settled on an approach to the change which is different from the original proposal. Please check the [updated issue](https://github.com/habitat-sh/core-packages/issues/56) for the details. In short the new approach would not require any header macro magic, it would rely on either standard library interposition over a shared system C library, or a wrapped function using the GNU ld's [`--wrap`](https://drewdevault.com/2016/07/19/Using-Wl-wrap-for-mocking-in-C.html) option.

Some considerations to address the concerns you mentioned around such a change:
- We aren't going to override the `dlopen` / `LoadLibrary` function, we would be wrapping it. Hence whatever the underlying behavior would remain untouched. Even if the function has different behavior on different UNIXes, our wrapper would still work.
- There is no need to track upstream as we delegate to the underlying system implementation. There is a problem if the external API of the function changes. But `dlopen` is part of the POSIX standard, and it hasn't been modified since a very long time. `LoadLibrary` has also not changed for a long time. Such changes would break a lot of software, not just Habitat.
- Nix does not have a fundamental fix for dynamic loading either, they rely on LD_LIBRARY_PATH and upstream modifications of language runtimes / libraries. We may not have the kind of resources Nix has to effect such changes.

I agree that clever solutions have a high knowledge and maintenance overhead but this is true of most of Habitat's design. It is essential complexity in the solution space of providing what Habitat provides. We already wrap compilers with specific flags, arrange system headers in specific orders, patch compilers, and generally do a lot of stuff under the hood to maintain the appearance of a regular build system to the plan author. 

In that context, this change is simply the addition of another flag that links in a library before the system C library
This can be documented in the wrappers, the gcc plan and in [other documentation](https://github.com/habitat-sh/core-packages/blob/main/docs/bootstrapping-gcc.md) we have around how we wrap gcc to make it work within Habitat.

In the same line of thought, the failure to fundamentally solve dynamic loading at the appropriate layer leads to such cryptic solutions as the one suggested in this plan. I view that failure to abstract this complexity as the bug here. The only reason rpath behavior optimization or lack of it matters here is because of that failure to handle dynamic loading.

I would also regard the superfluous rpaths added to every library/binary in the package, regardless of whether that binary/library actually needs it, as a bug. It just doesn't happen to have a noticeable enough effect for most users. However certain compilation processes do break because of it. Specifically, the compilation of the runtime linker in glibc breaks due to the addition of an rpath to the linker which it will crash on.

**Comment by jaymzh** [comment]
Yeah that seems reasonable to me.

**Comment by jaymzh** [comment]
OK so alllll that said - I believe this PR is no longer needed, correct? The backwards-compatible ldwrapper script should solve this without having to manually rebuild this specific gem, so I think we can close those now?

**Comment by atrniv** [comment]
@jaymzh , yes, this PR is not needed at this time. @smandaka , you can close it

**Pull request closed without merging**